corehq/form_processor/utils/general.py

```diff
@@ -9,3 +12,35 @@ def should_use_sql_backend(domain):
     # uncomment when ready for production
     # return USE_SQL_BACKEND.enabled(domain)
     return False
+
+
+def cache_return_value(fn):
+    """
+    Docorator to save the return value of a method or property.
+    Use this where you can't use quickcache or memoize because you can't
+    guarantee a consistent or unique lookup key e.g. forms that get deprecated
```

Comment:
The issue is that `memoize` uses a `dict` as the cache and the key is a tuple made up of the function args. This mechanism relies on the `hash` of the args and Django overrides the `__hash__` function to return `hash(self.primary_key)`. This means that you can't hash an unsaved model. That means if you try and access a memoized function on an unsaved model you get an exception: https://github.com/django/django/blob/master/django/db/models/base.py#L483

To get around this we were overriding the `__hash__` and using the `form_id` property as the hash value. This works fine except for the case where the `form_id` changes which happens when you deprecate a form. In this case any memoized properties you accessed prior to changing the ID will be re-fetched / calculated. But you can't do that with attachments because they are linked to the old ID so you'd have to query using the old ID and not the new one.

This seemed like the best solution to the problem, just save the value to a property on the object. The pattern is essentially this:

```python
def get_attachments(self)
    cached_attachments = getattr(self, '__cached_attachments', None)
    if cached_attachments is not None:
        return cached_attachments

    cached_attachments = fetch_from_db(self)
    self.__cached_attachments = cached_attachments
    return cached_attachments
```